File: Sources/BraintreeCore/BTHTTP.m
#import "BTHTTP.h"
#import "Braintree-Version.h"
#import "BTAPIPinnedCertificates.h"
#import "BTLogger_Internal.h"
#import "BTCacheDateValidator_Internal.h"
#include <sys/sysctl.h>

#if __has_include(<Braintree/BraintreeCore.h>)
#import <Braintree/BTClientToken.h>
#import <Braintree/BTHTTPErrors.h>
#import <Braintree/BTJSON.h>
#import <Braintree/BTURLUtils.h>
#else
#import <BraintreeCore/BTClientToken.h>
#import <BraintreeCore/BTHTTPErrors.h>
#import <BraintreeCore/BTJSON.h>
#import <BraintreeCore/BTURLUtils.h>
#endif

@interface BTHTTP () <NSURLSessionDelegate>

@property (nonatomic, strong) NSURL *baseURL;
@property (nonatomic, copy) NSString *authorizationFingerprint;
@property (nonatomic, copy) NSString *tokenizationKey;

@end

@implementation BTHTTP

- (instancetype)init {
    return nil;
}

- (instancetype)initWithBaseURL:(NSURL *)URL {
    self = [super init];
    if (self) {
        self.baseURL = URL;
        self.cacheDateValidator = [[BTCacheDateValidator alloc] init];
    }
    
    return self;
}

- (instancetype)initWithBaseURL:(NSURL *)URL authorizationFingerprint:(NSString *)authorizationFingerprint {
    self = [self initWithBaseURL:URL];
    if (self) {
        NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration ephemeralSessionConfiguration];
        configuration.HTTPAdditionalHeaders = self.defaultHeaders;

        NSOperationQueue *delegateQueue = [[NSOperationQueue alloc] init];
        delegateQueue.name = @"com.braintreepayments.BTHTTP";
        delegateQueue.maxConcurrentOperationCount = NSOperationQueueDefaultMaxConcurrentOperationCount;

        self.authorizationFingerprint = authorizationFingerprint;
        self.session = [NSURLSession sessionWithConfiguration:configuration delegate:self delegateQueue:delegateQueue];
        self.pinnedCertificates = [BTAPIPinnedCertificates trustedCertificates];
    }
    return self;
}

- (instancetype)initWithBaseURL:(nonnull NSURL *)URL tokenizationKey:(nonnull NSString *)tokenizationKey {
    if (self = [self initWithBaseURL:URL]) {
        NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration ephemeralSessionConfiguration];
        configuration.HTTPAdditionalHeaders = self.defaultHeaders;

        NSOperationQueue *delegateQueue = [[NSOperationQueue alloc] init];
        delegateQueue.name = @"com.braintreepayments.BTHTTP";
        delegateQueue.maxConcurrentOperationCount = NSOperationQueueDefaultMaxConcurrentOperationCount;

        self.session = [NSURLSession sessionWithConfiguration:configuration delegate:self delegateQueue:delegateQueue];
        self.pinnedCertificates = [BTAPIPinnedCertificates trustedCertificates];
        self.tokenizationKey = tokenizationKey;
    }
    return self;
}

- (instancetype)initWithClientToken:(BTClientToken *)clientToken {
    return [self initWithBaseURL:[clientToken.json[@"clientApiUrl"] asURL] authorizationFingerprint:clientToken.authorizationFingerprint];
}

- (instancetype)copyWithZone:(NSZone *)zone {
    BTHTTP *copiedHTTP;
    if (self.authorizationFingerprint) {
        copiedHTTP = [[[self class] allocWithZone:zone] initWithBaseURL:self.baseURL authorizationFingerprint:self.authorizationFingerprint];
    } else {
        copiedHTTP = [[[self class] allocWithZone:zone] initWithBaseURL:self.baseURL tokenizationKey:self.tokenizationKey];
    }
    
    copiedHTTP.pinnedCertificates = [_pinnedCertificates copy];
    return copiedHTTP;
}

- (void)setSession:(NSURLSession *)session {
    if (_session) {
        // If we already have a session, we need to invalidate it so that the session delegate is released to prevent a retain cycle
        [_session invalidateAndCancel];
    }

    _session = session;
}

#pragma mark - HTTP Methods

- (void)GET:(NSString *)aPath completion:(void(^)(BTJSON *body, NSHTTPURLResponse *response, NSError *error))completionBlock {
    [self GET:aPath parameters:nil completion:completionBlock];
}

- (void)GET:(NSString *)aPath parameters:(NSDictionary *)parameters shouldCache:(BOOL)shouldCache completion:(void(^)(BTJSON *body, NSHTTPURLResponse *response, NSError *error))completionBlock {
    if (shouldCache) {
        [self httpRequestWithCaching:@"GET" path:aPath parameters:parameters completion:completionBlock];
    } else {
        [self httpRequest:@"GET" path:aPath parameters:parameters completion:completionBlock];
    }
}

- (void)GET:(NSString *)aPath parameters:(NSDictionary *)parameters completion:(void(^)(BTJSON *body, NSHTTPURLResponse *response, NSError *error))completionBlock {
    [self httpRequest:@"GET" path:aPath parameters:parameters completion:completionBlock];
}

- (void)POST:(NSString *)aPath completion:(void(^)(BTJSON *body, NSHTTPURLResponse *response, NSError *error))completionBlock {
    [self POST:aPath parameters:nil completion:completionBlock];
}

- (void)POST:(NSString *)aPath parameters:(NSDictionary *)parameters completion:(void(^)(BTJSON *body, NSHTTPURLResponse *response, NSError *error))completionBlock {
    [self httpRequest:@"POST" path:aPath parameters:parameters completion:completionBlock];
}

- (void)PUT:(NSString *)aPath completion:(void(^)(BTJSON *body, NSHTTPURLResponse *response, NSError *error))completionBlock {
    [self PUT:aPath parameters:nil completion:completionBlock];
}

- (void)PUT:(NSString *)aPath parameters:(NSDictionary *)parameters completion:(void(^)(BTJSON *body, NSHTTPURLResponse *response, NSError *error))completionBlock {
    [self httpRequest:@"PUT" path:aPath parameters:parameters completion:completionBlock];
}

- (void)DELETE:(NSString *)aPath completion:(void(^)(BTJSON *body, NSHTTPURLResponse *response, NSError *error))completionBlock {
    [self DELETE:aPath parameters:nil completion:completionBlock];
}

- (void)DELETE:(NSString *)aPath parameters:(NSDictionary *)parameters completion:(void(^)(BTJSON *body, NSHTTPURLResponse *response, NSError *error))completionBlock {
    [self httpRequest:@"DELETE" path:aPath parameters:parameters completion:completionBlock];
}

#pragma mark - Underlying HTTP

- (void)httpRequestWithCaching:(NSString *)method path:(NSString *)aPath parameters:(NSDictionary *)parameters completion:(void(^)(BTJSON *body, NSHTTPURLResponse *response, NSError *error))completionBlock {
    [self createRequest:method path:aPath parameters:parameters completion:^(NSURLRequest *request, NSError *error) {
        if (error != nil) {
            [self handleRequestCompletion:nil request:nil shouldCache:NO response:nil error:error completionBlock:completionBlock];
            return;
        }
        
        NSCachedURLResponse *cachedResponse = [[NSURLCache sharedURLCache] cachedResponseForRequest:request];
        
        if ([self.cacheDateValidator isCacheInvalid:cachedResponse]) {
            [[NSURLCache sharedURLCache] removeAllCachedResponses];
            cachedResponse = nil;
        }
        
        // The increase in speed of API calls with cached configuration caused an increase in "network connection lost" errors.
        // Adding this delay allows us to throttle the network requests slightly to reduce load on the servers and decrease connection lost errors.
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.1 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
            if (cachedResponse != nil) {
                [self handleRequestCompletion:cachedResponse.data request:nil shouldCache:NO response:cachedResponse.response error:nil completionBlock:completionBlock];
            } else {
                NSURLSessionTask *task = [self.session dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                    [self handleRequestCompletion:data request:request shouldCache:YES response:response error:error completionBlock:completionBlock];
                }];
                [task resume];
            }
        });
    }];
}

- (void)httpRequest:(NSString *)method path:(NSString *)aPath parameters:(NSDictionary *)parameters completion:(void(^)(BTJSON *body, NSHTTPURLResponse *response, NSError *error))completionBlock {
    [self createRequest:method path:aPath parameters:parameters completion:^(NSURLRequest *request, NSError *error) {
        if (error != nil) {
            [self handleRequestCompletion:nil request:nil shouldCache:NO response:nil error:error completionBlock:completionBlock];
            return;
        }
        NSURLSessionTask *task = [self.session dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
            [self handleRequestCompletion:data request:request shouldCache:NO response:response error:error completionBlock:completionBlock];
        }];
        [task resume];
    }];
}

- (void)createRequest:(NSString *)method path:(NSString *)aPath parameters:(NSDictionary *)parameters completion:(void(^)(NSURLRequest *request, NSError *error))completionBlock {
    BOOL hasHttpPrefix = aPath != nil && [aPath hasPrefix:@"http"];
    if (!hasHttpPrefix && (!self.baseURL || [self.baseURL.absoluteString isEqualToString:@""])) {
        NSMutableDictionary *errorUserInfo = [NSMutableDictionary new];
        if (method) errorUserInfo[@"method"] = method;
        if (aPath) errorUserInfo[@"path"] = aPath;
        if (parameters) errorUserInfo[@"parameters"] = parameters;
        completionBlock(nil, [NSError errorWithDomain:BTHTTPErrorDomain code:BTHTTPErrorCodeMissingBaseURL userInfo:errorUserInfo]);
        return;
    }

    BOOL isNotDataURL = ![self.baseURL.scheme isEqualToString:@"data"];
    NSURL *fullPathURL;
    if (aPath && isNotDataURL) {
        if (hasHttpPrefix) {
            fullPathURL = [NSURL URLWithString:aPath];
        } else {
            fullPathURL = [self.baseURL URLByAppendingPathComponent:aPath];
        }
    } else {
        fullPathURL = self.baseURL;
    }

    if (parameters == nil) {
        parameters = [NSDictionary dictionary];
    }
    NSMutableDictionary *mutableParameters = [NSMutableDictionary dictionaryWithDictionary:parameters];
    if (self.authorizationFingerprint) {
        mutableParameters[@"authorization_fingerprint"] = self.authorizationFingerprint;
    }
    parameters = [mutableParameters copy];

    if (!fullPathURL) {
        // baseURL can be non-nil (e.g. an empty string) and still return nil for -URLByAppendingPathComponent:
        // causing a crash when NSURLComponents.componentsWithString is called with nil.
        NSMutableDictionary *errorUserInfo = [NSMutableDictionary new];
        if (method) errorUserInfo[@"method"] = method;
        if (aPath) errorUserInfo[@"path"] = aPath;
        if (parameters) errorUserInfo[@"parameters"] = parameters;
        errorUserInfo[NSLocalizedFailureReasonErrorKey] = @"fullPathURL was nil";
        completionBlock(nil, [NSError errorWithDomain:BTHTTPErrorDomain code:BTHTTPErrorCodeMissingBaseURL userInfo:errorUserInfo]);
        return;
    }
    
    NSURLComponents *components = [NSURLComponents componentsWithString:fullPathURL.absoluteString];

    NSMutableDictionary *headers = [NSMutableDictionary dictionaryWithDictionary:self.defaultHeaders];

    NSMutableURLRequest *request;

    if ([method isEqualToString:@"GET"] || [method isEqualToString:@"DELETE"]) {
        if (isNotDataURL) {
            components.percentEncodedQuery = [BTURLUtils queryStringWithDictionary:parameters];
        }
        request = [NSMutableURLRequest requestWithURL:components.URL];
    } else {
        request = [NSMutableURLRequest requestWithURL:components.URL];

        NSError *jsonSerializationError;
        NSData *bodyData;

        if ([parameters isKindOfClass:[NSDictionary class]]) {
            bodyData = [NSJSONSerialization dataWithJSONObject:parameters
                                                       options:0
                                                         error:&jsonSerializationError];
        }

        if (jsonSerializationError != nil) {
            completionBlock(nil, jsonSerializationError);
            return;
        }

        [request setHTTPBody:bodyData];
        headers[@"Content-Type"]  = @"application/json; charset=utf-8";
    }
    if (self.tokenizationKey) {
        headers[@"Client-Key"] = self.tokenizationKey;
    }
    [request setAllHTTPHeaderFields:headers];

    [request setHTTPMethod:method];
    
    completionBlock(request, nil);
}

- (void)handleRequestCompletion:(NSData *)data request:(NSURLRequest *)request shouldCache:(BOOL)shouldCache response:(NSURLResponse *)response error:(NSError *)error completionBlock:(void(^)(BTJSON *body, NSHTTPURLResponse *response, NSError *error))completionBlock {
    // Handle errors for which the response is irrelevant
    // e.g. SSL, unavailable network, etc.
    if (error != nil) {
        [self callCompletionBlock:completionBlock body:nil response:nil error:error];
        return;
    }

    NSHTTPURLResponse *httpResponse;
    if ([response isKindOfClass:[NSHTTPURLResponse class]]) {
        httpResponse = (NSHTTPURLResponse *)response;
    } else if ([response.URL.scheme isEqualToString:@"data"]) {
        httpResponse = [[NSHTTPURLResponse alloc] initWithURL:response.URL statusCode:200 HTTPVersion:nil headerFields:nil];
    }

    NSString *responseContentType = [response MIMEType];

    NSMutableDictionary *errorUserInfo = [NSMutableDictionary new];
    errorUserInfo[BTHTTPURLResponseKey] = httpResponse;

    if (httpResponse.statusCode >= 400) {
        errorUserInfo[NSLocalizedFailureReasonErrorKey] = [NSHTTPURLResponse localizedStringForStatusCode:httpResponse.statusCode];

        BTJSON *json;
        if ([responseContentType isEqualToString:@"application/json"]) {
            json = (data.length == 0) ? [BTJSON new] : [[BTJSON alloc] initWithData:data];
            if (!json.isError) {
                errorUserInfo[BTHTTPJSONResponseBodyKey] = json;
                NSString *errorResponseMessage = [json[@"error"][@"developer_message"] isString] ? [json[@"error"][@"developer_message"] asString] : [json[@"error"][@"message"] asString];
                if (errorResponseMessage) {
                    errorUserInfo[NSLocalizedDescriptionKey] = errorResponseMessage;
                }
            }
        }
        
        BTHTTPErrorCode errorCode = httpResponse.statusCode >= 500 ? BTHTTPErrorCodeServerError : BTHTTPErrorCodeClientError;
        if (httpResponse.statusCode == 429) {
            errorCode = BTHTTPErrorCodeRateLimitError;
            errorUserInfo[NSLocalizedDescriptionKey] = @"You are being rate-limited.";
            errorUserInfo[NSLocalizedRecoverySuggestionErrorKey] = @"Please try again in a few minutes.";
        } else if (httpResponse.statusCode >= 500) {
            errorUserInfo[NSLocalizedRecoverySuggestionErrorKey] = @"Please try again later.";
        }
        
        NSError *error = [NSError errorWithDomain:BTHTTPErrorDomain
                                                     code:errorCode
                                                 userInfo:[errorUserInfo copy]];
        [self callCompletionBlock:completionBlock body:json response:httpResponse error:error];
        return;
    }

    // Empty response is valid
    BTJSON *json = (data.length == 0) ? [BTJSON new] : [[BTJSON alloc] initWithData:data];
    if (json.isError) {
        if (![responseContentType isEqualToString:@"application/json"]) {
            // Return error for unsupported response type
            errorUserInfo[NSLocalizedFailureReasonErrorKey] = [NSString stringWithFormat:@"BTHTTP only supports application/json responses, received Content-Type: %@", responseContentType];
            NSError *returnedError = [NSError errorWithDomain:BTHTTPErrorDomain
                                                         code:BTHTTPErrorCodeResponseContentTypeNotAcceptable
                                                     userInfo:[errorUserInfo copy]];
            [self callCompletionBlock:completionBlock body:nil response:nil error:returnedError];
        } else {
            [self callCompletionBlock:completionBlock body:nil response:nil error:json.asError];
        }
        return;
    }
    
    // We should only cache the response if we do not have an error and status code is 2xx
    BOOL successStatusCode = httpResponse.statusCode >= 200 && httpResponse.statusCode < 300;
    if (request != nil && shouldCache && successStatusCode) {
        NSCachedURLResponse *cachedURLResponse = [[NSCachedURLResponse alloc]initWithResponse:response data:data];
        [[NSURLCache sharedURLCache] storeCachedResponse:cachedURLResponse forRequest:request];
    }

    [self callCompletionBlock:completionBlock body:json response:httpResponse error:nil];
}

- (void)callCompletionBlock:(void(^)(BTJSON *body, NSHTTPURLResponse *response, NSError *error))completionBlock
                       body:(BTJSON *)jsonBody
                   response:(NSHTTPURLResponse *)response
                      error:(NSError *)error {
    if (completionBlock) {
        dispatch_async(self.dispatchQueue, ^{
            completionBlock(jsonBody, response, error);
        });
    }
}

- (dispatch_queue_t)dispatchQueue {
    return _dispatchQueue ?: dispatch_get_main_queue();
}

#pragma mark - Default Headers

- (NSDictionary *)defaultHeaders {
    return @{ @"User-Agent": [self userAgentString],
              @"Accept": [self acceptString],
              @"Accept-Language": [self acceptLanguageString] };
}

- (NSString *)userAgentString {
    return [NSString stringWithFormat:@"Braintree/iOS/%@", BRAINTREE_VERSION];
}

- (NSString *)platformString {
    size_t size = 128;
    char *hwModel = alloca(size);

    if (sysctlbyname("hw.model", hwModel, &size, NULL, 0) != 0) {
        return nil;
    }

    NSString *hwModelString = [NSString stringWithCString:hwModel encoding:NSUTF8StringEncoding];
#if TARGET_IPHONE_SIMULATOR
    hwModelString = [hwModelString stringByAppendingString:@"(simulator)"];
#endif
    return hwModelString;
}

- (NSString *)architectureString {
    size_t size = 128;
    char *hwMachine = alloca(size);

    if (sysctlbyname("hw.machine", hwMachine, &size, NULL, 0) != 0) {
        return nil;
    }

    return [NSString stringWithCString:hwMachine encoding:NSUTF8StringEncoding];
}

- (NSString *)acceptString {
    return @"application/json";
}

- (NSString *)acceptLanguageString {
    NSLocale *locale = [NSLocale currentLocale];
    return [NSString stringWithFormat:@"%@-%@",
            [locale objectForKey:NSLocaleLanguageCode],
            [locale objectForKey:NSLocaleCountryCode]];
}

#pragma mark - Helpers

- (NSArray *)pinnedCertificateData {
    NSMutableArray *pinnedCertificates = [NSMutableArray array];
    for (NSData *certificateData in self.pinnedCertificates) {
        [pinnedCertificates addObject:(__bridge_transfer id)SecCertificateCreateWithData(NULL, (__bridge CFDataRef)certificateData)];
    }
    return pinnedCertificates;
}

- (void)URLSession:(__unused NSURLSession *)session didReceiveChallenge:(NSURLAuthenticationChallenge *)challenge completionHandler:(void (^)(NSURLSessionAuthChallengeDisposition, NSURLCredential *))completionHandler {
    if ([[[challenge protectionSpace] authenticationMethod] isEqualToString:NSURLAuthenticationMethodServerTrust]) {
        NSString *domain = challenge.protectionSpace.host;
        SecTrustRef serverTrust = [[challenge protectionSpace] serverTrust];

        NSArray *policies = @[(__bridge_transfer id)SecPolicyCreateSSL(true, (__bridge CFStringRef)domain)];
        SecTrustSetPolicies(serverTrust, (__bridge CFArrayRef)policies);
        SecTrustSetAnchorCertificates(serverTrust, (__bridge CFArrayRef)self.pinnedCertificateData);

        CFErrorRef error;
        BOOL trusted = SecTrustEvaluateWithError(serverTrust, &error);

        if (trusted && error == nil) {
            NSURLCredential *credential = [NSURLCredential credentialForTrust:serverTrust];
            completionHandler(NSURLSessionAuthChallengeUseCredential, credential);
        } else {
            completionHandler(NSURLSessionAuthChallengeRejectProtectionSpace, NULL);
        }
    } else {
        completionHandler(NSURLSessionAuthChallengePerformDefaultHandling, NULL);
    }
}

- (BOOL)isEqualToHTTP:(BTHTTP *)http {
    return [self.baseURL isEqual:http.baseURL] && [self.authorizationFingerprint isEqualToString:http.authorizationFingerprint];
}

- (BOOL)isEqual:(id)object {
    if (self == object) {
        return YES;
    }

    if ([object isKindOfClass:[BTHTTP class]]) {
        return [self isEqualToHTTP:object];
    }

    return NO;
}

@end

File: Sources/BraintreeCore/BTHTTP.swift
import Foundation
import Security

// swiftlint:disable type_body_length file_length
/// Performs HTTP methods on the Braintree Client API
class BTHTTP: NSObject, URLSessionTaskDelegate {

    typealias RequestCompletion = (BTJSON?, HTTPURLResponse?, Error?) -> Void

    // MARK: - Internal Properties

    /// An array of pinned certificates, each a Data instance consisting of DER encoded x509 certificates
    let pinnedCertificates: [Data] = BTAPIPinnedCertificates.trustedCertificates()

    /// DispatchQueue on which asynchronous code will be executed. Defaults to `DispatchQueue.main`.
    var dispatchQueue = DispatchQueue.main
    
    /// A URL set to override the URLs derived from the ClientAuthorization or BTConfiguration response
    let customBaseURL: URL?
    
    let authorization: ClientAuthorization
    
    weak var networkTimingDelegate: BTHTTPNetworkTiming?

    /// Session exposed for testing
    lazy var session: URLSession = {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.httpAdditionalHeaders = defaultHeaders
        configuration.timeoutIntervalForRequest = 30
        configuration.timeoutIntervalForResource = 30
        
        let delegateQueue = OperationQueue()
        delegateQueue.name = "com.braintreepayments.BTHTTP"
        
        return URLSession(configuration: configuration, delegate: self, delegateQueue: delegateQueue)
    }()

    var defaultHeaders: [String: String] {
        [
            "User-Agent": userAgentString,
            "Accept": acceptString,
            "Accept-Language": acceptLanguageString
        ]
    }

    var userAgentString: String {
        "Braintree/iOS/\(BTCoreConstants.braintreeSDKVersion)"
    }

    var acceptString: String {
        "application/json"
    }

    var acceptLanguageString: String {
        "\(Locale.current.languageCode ?? "en")-\(Locale.current.regionCode ?? "US")"
    }
    
    // MARK: - Internal Initializers
    
    /// Initialize `BTHTTP` with an Authorization credential
    /// - Parameters:
    ///   - authorization: a CilentToken or TokenizationKey
    ///   - customBaseURL: an optional baseURL override
    required init(authorization: ClientAuthorization, customBaseURL: URL? = nil) {
        self.authorization = authorization
        self.customBaseURL = customBaseURL
    }

    // MARK: - HTTP Methods

    func get(_ path: String, configuration: BTConfiguration? = nil, parameters: Encodable? = nil, completion: @escaping RequestCompletion) {
        httpRequest(method: .get, path: path, configuration: configuration, parameters: parameters, completion: completion)
    }
    
    func get(
        _ path: String,
        configuration: BTConfiguration? = nil,
        parameters: Encodable? = nil
    ) async throws -> (BTJSON?, HTTPURLResponse?) {
        try await httpRequest(
            method: .get,
            path: path,
            configuration: configuration,
            parameters: parameters
        )
    }
    
    func post(
        _ path: String,
        configuration: BTConfiguration? = nil,
        parameters: Encodable? = nil,
        headers: [String: String]? = nil,
        completion: @escaping RequestCompletion
    ) {
        if authorization.type == .invalidAuthorization {
            completion(nil, nil, BTAPIClientError.invalidAuthorization(authorization.originalValue))
            return
        }
        
        httpRequest(
            method: .post,
            path: path,
            configuration: configuration,
            parameters: parameters,
            headers: headers,
            completion: completion
        )
    }
    
    func post(
        _ path: String,
        configuration: BTConfiguration? = nil,
        parameters: Encodable? = nil,
        headers: [String: String]? = nil
    ) async throws -> (BTJSON?, HTTPURLResponse?) {
        try await httpRequest(
            method: .post,
            path: path,
            configuration: configuration,
            parameters: parameters,
            headers: headers
        )
    }

    // MARK: - HTTP Method Helpers

    // TODO: Remove once code calling completion handler version is removed.
    func httpRequest(
        method: BTHTTPMethod,
        path: String,
        configuration: BTConfiguration? = nil,
        parameters: Encodable? = nil,
        headers: [String: String]? = nil,
        completion: RequestCompletion?
    ) {
        guard let completion else { return }

        Task { [weak self] in
            guard let self else {
                DispatchQueue.main.async {
                    completion(nil, nil, BTHTTPError.deallocated("BTHTTP"))
                }
                return
            }

            do {
                let (json, httpResponse) = try await httpRequest(
                    method: method,
                    path: path,
                    configuration: configuration,
                    parameters: parameters,
                    headers: headers
                )
                let jsonDict = json.asDictionary() as? [String: Any]
                let jsonValue = json.value
                let capturedResponse = httpResponse
                self.dispatchQueue.async {
                    let reconstructedJSON = jsonDict.map { BTJSON(value: $0) } ?? BTJSON(value: jsonValue)
                    completion(reconstructedJSON, capturedResponse, nil)
                }
            } catch {
                let bodyFromError = (error as NSError).userInfo[BTCoreConstants.jsonResponseBodyKey] as? BTJSON
                let bodyDict = bodyFromError?.asDictionary() as? [String: Any]
                let response = (error as NSError).userInfo[BTCoreConstants.urlResponseKey] as? HTTPURLResponse
                let capturedError = error

                self.dispatchQueue.async {
                    let reconstructedBody = bodyDict.map { BTJSON(value: $0) }
                    completion(reconstructedBody, response, capturedError)
                }
            }
        }
    }
    
    func httpRequest(
        method: BTHTTPMethod,
        path: String,
        configuration: BTConfiguration? = nil,
        parameters: Encodable? = nil,
        headers: [String: String]? = nil
    ) async throws -> (BTJSON, HTTPURLResponse) {
        let request = try createRequest(
            method: method,
            path: path,
            configuration: configuration,
            parameters: parameters,
            headers: headers
        )
        let (data, response) = try await session.data(for: request)
        return try handleRequestCompletion(data: data, request: request, response: response)
    }

    func createRequest(
        method: BTHTTPMethod,
        path: String,
        configuration: BTConfiguration? = nil,
        parameters: Encodable? = nil,
        headers: [String: String]? = [:]
    ) throws -> URLRequest {
        var fullPathURL: URL
        if let customBaseURL {
            fullPathURL = customBaseURL.appendingPathComponent(path)
        } else {
            fullPathURL = configuration?.clientAPIURL?.appendingPathComponent(path) ?? authorization.configURL
        }

        if fullPathURL.absoluteString.isEmpty {
            var errorUserInfo: [String: Any] = [:]
            errorUserInfo["method"] = method.rawValue
            errorUserInfo["path"] = path
            errorUserInfo["parameters"] = parameters

            throw BTHTTPError.missingBaseURL(errorUserInfo)
        }
        
        var mutableParameters = NSMutableDictionary()
        if let parameters {
            let encodedData = try JSONEncoder().encode(parameters)
            if let jsonObject = try JSONSerialization.jsonObject(with: encodedData) as? [String: Any] {
                mutableParameters = NSMutableDictionary(dictionary: jsonObject)
            }
        }
        
        // TODO: - Investigate for parity on JS and Android
        // JIRA - DTBTSDK-2682
        if authorization.type == .clientToken, !fullPathURL.isPayPalURL {
            mutableParameters["authorization_fingerprint"] = authorization.bearer
        }

        return try buildHTTPRequest(
            method: method,
            url: fullPathURL,
            parameters: mutableParameters,
            headers: headers
        )
    }

    func buildHTTPRequest(
        method: BTHTTPMethod,
        url: URL,
        parameters: NSMutableDictionary? = [:],
        headers additionalHeaders: [String: String]? = nil
    ) throws -> URLRequest {
        guard var components = URLComponents(string: url.absoluteString) else {
            throw BTHTTPError.urlStringInvalid
        }
        
        var headers: [String: String] = defaultHeaders
        var request: URLRequest
        
        if url.isPayPalURL {
            headers = [:]
            if authorization.type == .clientToken {
                headers["Authorization"] = "Bearer \(authorization.bearer)"
            }
        } else {
            headers = defaultHeaders
            if authorization.type == .tokenizationKey {
                headers["Client-Key"] = authorization.bearer
            }
        }
        
        if let additionalHeaders {
            headers = headers.merging(additionalHeaders) { $1 }
        }
        
        if method == .get || method == .delete {
            components.percentEncodedQuery = BTURLUtils.queryString(from: parameters ?? [:])
            
            guard let urlFromComponents = components.url else {
                throw BTHTTPError.urlStringInvalid
            }
            
            request = URLRequest(url: urlFromComponents)
        } else {
            guard let urlFromComponents = components.url else {
                throw BTHTTPError.urlStringInvalid
            }
            
            request = URLRequest(url: urlFromComponents)
            
            var bodyData: Data
            
            do {
                bodyData = try JSONSerialization.data(withJSONObject: parameters ?? [:])
            } catch {
                throw error
            }
            
            request.httpBody = bodyData
            headers["Content-Type"] = "application/json; charset=utf-8"
        }
        
        if authorization.type == .tokenizationKey {
            headers["Client-Key"] = authorization.originalValue
        }
        
        request.allHTTPHeaderFields = headers
        request.httpMethod = method.rawValue
        
        return request
    }

    func handleRequestCompletion(
        data: Data? = nil,
        request: URLRequest? = nil,
        response: URLResponse? = nil,
        error: Error? = nil
    ) throws -> (BTJSON, HTTPURLResponse) {
        if let error {
            throw error
        }
        
        guard let response, let httpResponse = createHTTPResponse(response: response) else {
            throw BTHTTPError.httpResponseInvalid
        }
        
        guard let data else {
            throw BTHTTPError.dataNotFound
        }
        
        if httpResponse.statusCode >= 400 {
            let (_, error) = try handleHTTPResponseError(response: httpResponse, data: data)
            throw error
        }
        
        let json: BTJSON = data.isEmpty ? BTJSON() : BTJSON(data: data)
        if json.isError {
            try handleJSONResponseError(json: json, response: response)
        }
        
        return (json, httpResponse)
    }

    func createHTTPResponse(response: URLResponse) -> HTTPURLResponse? {
        if let httpResponse = response as? HTTPURLResponse {
            return httpResponse
        } else if let url = response.url, url.scheme == "data" {
            return HTTPURLResponse(
                url: url,
                statusCode: 200,
                httpVersion: nil,
                headerFields: nil
            )
        }
        return nil
    }

    func handleHTTPResponseError(response: HTTPURLResponse, data: Data) throws -> (BTJSON, Error) {
        let responseContentType: String? = response.mimeType
        var errorUserInfo: [String: Any] = [BTCoreConstants.urlResponseKey: response]
        
        errorUserInfo[NSLocalizedFailureReasonErrorKey] = [HTTPURLResponse.localizedString(forStatusCode: response.statusCode)]
        
        var json = BTJSON()
        if responseContentType == "application/json" {
            json = data.isEmpty ? BTJSON() : BTJSON(data: data)
            if !json.isError {
                errorUserInfo[BTCoreConstants.jsonResponseBodyKey] = json
                let errorResponseMessage: String? = json["error"]["developer_message"].asString() ?? json["error"]["message"].asString()
                if errorResponseMessage != nil {
                    errorUserInfo[NSLocalizedDescriptionKey] = errorResponseMessage
                }
            }
        }
        
        var error = BTHTTPError.clientError(errorUserInfo)
        
        if response.statusCode == 429 {
            errorUserInfo[NSLocalizedDescriptionKey] = "You are being rate-limited."
            errorUserInfo[NSLocalizedRecoverySuggestionErrorKey] = "Please try again in a few minutes."
            error = BTHTTPError.rateLimitError(errorUserInfo)
        } else if response.statusCode <= 500 {
            error = BTHTTPError.clientError(errorUserInfo)
        } else if response.statusCode >= 500 {
            errorUserInfo[NSLocalizedRecoverySuggestionErrorKey] = "Please try again later."
            error = BTHTTPError.serverError(errorUserInfo)
        }
        
        return (json, error)
    }

    func handleJSONResponseError(
        json: BTJSON,
        response: URLResponse
    ) throws {
        let responseContentType: String? = response.mimeType
        var errorUserInfo: [String: Any] = [BTCoreConstants.urlResponseKey: response]

        if let contentType = responseContentType, contentType != "application/json" {
            // Return error for unsupported response type
            let message = "BTHTTP only supports application/json responses, received Content-Type: \(contentType)"
            errorUserInfo[NSLocalizedFailureReasonErrorKey] = message
            throw BTHTTPError.responseContentTypeNotAcceptable(errorUserInfo)
        } else if let error = json.asError() {
            throw error
        }
    }

    // MARK: - Helper functions

    func pinnedCertificateData() -> [SecCertificate] {
        var certificates: [SecCertificate] = []

        for certificateData in pinnedCertificates {
            guard let certificate = SecCertificateCreateWithData(nil, certificateData as CFData) else { return [] }
            certificates.append(certificate)
        }

        return certificates
    }

    // MARK: - URLSessionTaskDelegate conformance

    func urlSession(
        _ session: URLSession,
        didReceive challenge: URLAuthenticationChallenge,
        completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
    ) {
        if challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust {
            let domain: String = challenge.protectionSpace.host
            // swiftlint:disable force_unwrapping
            let serverTrust: SecTrust = challenge.protectionSpace.serverTrust!
            // swiftlint:enable force_unwrapping

            let policies: [SecPolicy] = [SecPolicyCreateSSL(true, domain as CFString)]
            SecTrustSetPolicies(serverTrust, policies as CFArray)
            SecTrustSetAnchorCertificates(serverTrust, pinnedCertificateData() as CFArray)

            var error: CFError?
            let trusted: Bool = SecTrustEvaluateWithError(serverTrust, &error)

            if trusted && error == nil {
                let credential = URLCredential(trust: serverTrust)
                completionHandler(.useCredential, credential)
            } else {
                completionHandler(.rejectProtectionSpace, nil)
            }
        } else {
            completionHandler(.performDefaultHandling, nil)
        }
    }

    func urlSession(_ session: URLSession, task: URLSessionTask, didFinishCollecting metrics: URLSessionTaskMetrics) {
        metrics.transactionMetrics.forEach { transaction in
            if let startDate = transaction.fetchStartDate,
                let endDate = transaction.responseEndDate,
                var path = transaction.request.url?.path {
                if path.contains("graphql"),
                    let data = task.originalRequest?.httpBody,
                    let mutationName = getGraphQLMutationName(data) {
                        path = mutationName
                }
                
                networkTimingDelegate?.fetchAPITiming(
                    path: path,
                    connectionStartTime: transaction.connectStartDate?.utcTimestampMilliseconds,
                    requestStartTime: transaction.requestStartDate?.utcTimestampMilliseconds,
                    startTime: startDate.utcTimestampMilliseconds,
                    endTime: endDate.utcTimestampMilliseconds
                )
            }
        }
    }
    
    private func getGraphQLMutationName(_ data: Data) -> String? {
        let json = try? JSONSerialization.jsonObject(with: data)
        let body = BTJSON(value: json)
        
        guard let query = body["query"].asString() else {
            return nil
        }

        let queryDiscardHolder = query.replacingOccurrences(of: #"^[^\(]*"#, with: "", options: .regularExpression)
        let finalQuery = query.replacingOccurrences(of: queryDiscardHolder, with: "")
        return finalQuery
    }
}
